File: juegos1000tres-backend/src/main/java/com/juegos1000tres/juegos1000tres_backend/comunicacion/Traductor.java
package com.juegos1000tres.juegos1000tres_backend.comunicacion;

import java.util.Objects;

import com.juegos1000tres.juegos1000tres_backend.modelos.Juego;

public class Traductor<PAYLOAD> {

    private final Conexion<PAYLOAD> conexion;
    private final Envio<PAYLOAD> envio;
    private final Recibo<PAYLOAD> recibo;
    private final Class<PAYLOAD> clasePayload;

    public Traductor(Conexion<PAYLOAD> conexion, Envio<PAYLOAD> envio, Recibo<PAYLOAD> recibo) {
        this.conexion = Objects.requireNonNull(conexion, "La conexion es obligatoria");
        this.envio = Objects.requireNonNull(envio, "La estrategia de envio es obligatoria");
        this.recibo = Objects.requireNonNull(recibo, "La estrategia de recibo es obligatoria");
        this.clasePayload = this.conexion.getClasePayload();

        validarCompatibilidadDePayload(this.clasePayload, this.envio.getClasePayload(), "Envio");
        validarCompatibilidadDePayload(this.clasePayload, this.recibo.getClasePayload(), "Recibo");
    }

    public PAYLOAD traducirEnviableAFormato(Enviable enviable) {
        return envio.traducirEnviableAFormato(enviable);
    }

    public <T extends Enviable> T traducirFormatoAEnviable(PAYLOAD payload, Class<T> tipoEnviable) {
        return recibo.traducirFormatoAEnviable(payload, tipoEnviable);
    }

    public void enviar(Enviable enviable) {
        PAYLOAD payload = traducirEnviableAFormato(enviable);
        conexion.enviar(payload);
    }

    public <T extends Enviable> T recibir(Class<T> tipoEnviable) {
        PAYLOAD payload = conexion.recibir();
        return traducirFormatoAEnviable(payload, tipoEnviable);
    }

    public <T extends Enviable> T recibirYNotificarJuego(Juego juego, Class<T> tipoEnviable) {
        Objects.requireNonNull(juego, "El juego es obligatorio");
        Objects.requireNonNull(tipoEnviable, "El tipo de mensaje es obligatorio");

        T mensaje = recibir(tipoEnviable);
        juego.procesarMensajeEntrante(mensaje);
        return mensaje;
    }

    public Class<PAYLOAD> getClasePayload() {
        return clasePayload;
    }

    private void validarCompatibilidadDePayload(Class<PAYLOAD> claseBase, Class<PAYLOAD> claseDependencia, String dependencia) {
        if (!Objects.equals(claseBase, claseDependencia)) {
            throw new IllegalArgumentException(
                    "La clase de payload de " + dependencia + " no coincide con la de Conexion");
        }
    }
}

File: juegos1000tres-frontend/src/comunicacion/Traductor.ts
import { Enviable, type EnviableConstructor } from "./Enviable.js";
import { Conexion } from "./Conexion.js";
import { Envio } from "./Envio.js";
import { Recibo } from "./Recibo.js";

type JuegoReceptor = {
  procesarMensajeEntrante(mensaje: Enviable): Promise<void> | void;
};

export class Traductor<PAYLOAD> {
  private readonly conexion: Conexion<PAYLOAD>;
  private readonly envio: Envio<PAYLOAD>;
  private readonly recibo: Recibo<PAYLOAD>;
  private readonly tipoPayload: string;

  constructor(conexion: Conexion<PAYLOAD>, envio: Envio<PAYLOAD>, recibo: Recibo<PAYLOAD>) {
    if (!conexion) {
      throw new Error("La conexion es obligatoria");
    }

    if (!envio) {
      throw new Error("La estrategia de envio es obligatoria");
    }

    if (!recibo) {
      throw new Error("La estrategia de recibo es obligatoria");
    }

    this.conexion = conexion;
    this.envio = envio;
    this.recibo = recibo;
    this.tipoPayload = this.conexion.getTipoPayload();

    this.validarCompatibilidadPayload();
  }

  traducirEnviableAFormato(enviable: Enviable): PAYLOAD {
    return this.envio.traducirEnviableAFormato(enviable);
  }

  traducirFormatoAEnviable<T extends Enviable>(
    payload: PAYLOAD,
    tipoEnviable: EnviableConstructor<T>
  ): T {
    return this.recibo.traducirFormatoAEnviable(payload, tipoEnviable);
  }

  async enviar(enviable: Enviable): Promise<void> {
    const payload = this.traducirEnviableAFormato(enviable);
    await this.conexion.enviar(payload);
  }

  async recibir<T extends Enviable>(tipoEnviable: EnviableConstructor<T>): Promise<T> {
    const payload = await this.conexion.recibir();
    return this.traducirFormatoAEnviable(payload, tipoEnviable);
  }

  async recibirYNotificarJuego<T extends Enviable>(
    juegoConexion: JuegoReceptor,
    tipoEnviable: EnviableConstructor<T>
  ): Promise<T> {
    if (!juegoConexion) {
      throw new Error("El juego es obligatorio");
    }

    const mensaje = await this.recibir(tipoEnviable);
    await juegoConexion.procesarMensajeEntrante(mensaje);
    return mensaje;
  }

  getTipoPayload(): string {
    return this.tipoPayload;
  }

  private validarCompatibilidadPayload(): void {
    const tipoEnvio = this.envio.getTipoPayload();
    const tipoRecibo = this.recibo.getTipoPayload();

    if (this.tipoPayload !== tipoEnvio) {
      throw new Error("El tipo de payload de Envio no coincide con Conexion");
    }

    if (this.tipoPayload !== tipoRecibo) {
      throw new Error("El tipo de payload de Recibo no coincide con Conexion");
    }
  }
}
